ppseq/model.py

```diff
@@ -11,6 +8,473 @@
+
+        Returns
+        -------
+        Scalar log probability
+        """
+        D = self.template_duration
+        kernel = torch.flip(self.templates.permute(1,0,2),[2])
+        background = self.get_background_rates(X)
+        rates = background + F.conv1d(amplitudes, kernel, padding=D-1)[:,:-D+1]
+        rates = torch.clamp(rates, min=1e-7)
+        if rows is None or cols is None:
+            return torch.sum(dist.Poisson(rates).log_prob(data))
+
+        data_selected = data[rows,:][:,cols]
+        rates_selected = rates[rows,:][:,cols]
+        poisson_dist = dist.Poisson(rates_selected)
+        log_probs = poisson_dist.log_prob(data_selected)
+        log_likelihood = torch.sum(log_probs)
+        return log_likelihood
+
+    def _update_amplitudes(self, data, amplitudes, X):
+        D, T = self.template_duration, data.shape[1]
+        W = self.templates
+        kernel = torch.flip(W.permute(1,0,2), [2])
+        background = self.get_background_rates(X)
+        rates = background + F.conv1d(amplitudes, kernel, padding=D-1)[:,:-D+1]
+        ratio = data / (rates + 1e-7)
+
+        alpha_post = amplitudes * F.conv1d(ratio, W, padding=D-1)[:,D-1:] + self.alpha_a0
+        beta_post = torch.sum(W, dim=(1,2)).unsqueeze(1).repeat(1,T) + self.beta_a0
+
+        # soft L1 penalty
+        if self.l1_amp > 0:
+            alpha_post -= self.l1_amp
+        
+        return torch.clip((alpha_post - 1) / beta_post, 0)
+
+    def create_smooth_covariates(self):
+        """Create smooth temporal features using Gaussian filtering or RBF basis functions"""
+        # Use slow RBF basis functions when data is not provided
+        if not hasattr(self, 'T_for_rbf'):
+            raise ValueError("Need to call create_smooth_covariates with data at least once, or set T_for_rbf")
+        
+        T = self.T_for_rbf
+        device = self.device
+        
+        # Create RBF basis functions with centers spread across time
+        n_basis = self.covariate_dim 
+        centers = torch.linspace(0, T-1, n_basis, device=device)
+        width = (T / (n_basis - 1)) * self.rbf_width  # Width scaled by rbf_width parameter
+        
+        # Create time indices
+        t = torch.arange(T, dtype=torch.float32, device=device)
+        
+        # Compute RBF basis functions
+        phi = []
+        for c in centers:
+            rbf = torch.exp(-(t - c)**2 / (2 * width**2))
+            phi.append(rbf)
+        
+        return torch.stack(phi)  # (covariate_dim, T)
+
+    def _update_base_rates(self, data, amplitudes, max_iter=10, tol=1e-4):
+        """
+        Vectorized Newton-Raphson update for GLM parameters
+        Updates all neurons simultaneously for speed
+        """
+        D = self.template_duration
+        T = data.shape[1]
+        N = self.num_neurons
+
+        # Store T for RBF basis functions
+        self.T_for_rbf = T
```

Comment:
If there is a better way to do T for RBF, it would be nice - if you want to do inference, it needs to be set again and can cause some issues... though model isn't currently set for inference. but we can maybe implement this in a cleaner way next time